Fixes LRP service deletion/restoration and adds e2e tests for LRP.

Weil0ng opened the pull request:
Signed-off-by: Weilong Cui <cuiwl@google.com>

Weil0ng committed:
Adds a basic e2e test for LRP.

Adds 2 test cases:
1. checks LRP should redirect traffic *only* to local backend.
2. checks removing LRP restores original service (that traffic goes to
both backends).

Signed-off-by: Weilong Cui <cuiwl@google.com>

aanm commented on test/k8sT/manifests/lrp-test.yaml:
```diff
@@ -0,0 +1,136 @@
+SrcIP=10.80.0.9 DstIP=10.80.0.2
```
leftover?

Weil0ng replied:
Done.

brb commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,93 @@ var _ = Describe("K8sServicesTest", func() {
 		})
 	})
 
+	Context("Checks local redirect policy", func() {
```
LRP depends on the host-reachable services feature which is enabled only on `net-next` and `4.19` CI runs. To skip the suite on other CI runs of other types, you can use `SkipContextIf(func () bool { return !helpers.RunsOnNetNextOr419Kernel() }, "Checks local redirect policy", func () { ...`.

Weil0ng replied:
Ah makes sense, thanks!

brb commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,93 @@ var _ = Describe("K8sServicesTest", func() {
+
+		BeforeAll(func() {
+			deploymentYAML = helpers.ManifestGet(kubectl.BasePath(), "lrp-test.yaml")
+			lrpServiceName = "lrp-demo-service"
```
Nit: I think this can be declared as a const before `BeforeAll(..)`, can't it?

Weil0ng replied:
Done.

brb commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,93 @@ var _ = Describe("K8sServicesTest", func() {
+				return ip
+			}
+			// Should never reach here
+			return ""
```
Maybe then it makes sense to add a failing assertion instead of returning the empty string?

Weil0ng replied:
I think the assertion is covered before the loop above, this is just to make compiler happy, but maybe it's a bad practice?

brb commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,93 @@ var _ = Describe("K8sServicesTest", func() {
+			ciliumPods, err := kubectl.GetCiliumPods()
+			Expect(err).To(BeNil(), "Cannot get cilium pods")
+			for _, pod := range ciliumPods {
+				service := kubectl.CiliumExecMustSucceed(context.TODO(), pod, "cilium service list", "Cannot retrieve services on cilium pod")
```
Shall we grep for `clusterIP` to make sure that that service is of the LR type?

Weil0ng replied:
This is after LRP has been deployed, hence the service type should be of string "LocalRedirect", are you suggesting we check service type *before* deploying LRP too?

brb replied:
I meant to grep for a particular service to see whether it's type has changed to LocalRedirect, i.e. `fmt.Sprintf("cilium service list | grep %s", clusterIP)`. Because right now we are checking whether any service is of the LocalRedirect type.

Weil0ng replied:
Ah I see, done.

brb commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,93 @@ var _ = Describe("K8sServicesTest", func() {
+			testCurlFromPods("id=app1", tftpSVCURL, 10, 0)
+			monitorCancel1()
+			monitorRes1.ExpectSuccess()
+			monitorRes1.ExpectContains(fmt.Sprintf("SrcIP=%s DstIP=%s", app1IP, be1IP))
```
Alternatively, you could check against pod name. Echoserver (both - tcp and udp) should return `Hostname: ...` which contains a pod name.

Weil0ng replied:
Thanks! Changed to check the returned pod name, I think that's better :)

brb reviewed (commented):
Thanks!

aditighag commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,93 @@ var _ = Describe("K8sServicesTest", func() {
+		getSingletonPodIP := func(label string) string {
+			podsIPs, err := kubectl.GetPodsIPs(helpers.DefaultNamespace, label)
+			Expect(err).To(BeNil(), "Cannot get pod IP for %s pod", label)
+			Expect(len(podsIPs)).To(Equal(1), "Got more than 1 ip for %s pod", label)
```
Just to confirm this is only meant to simply the test workflow, right? LRP manager does process multiple podIPs.

Weil0ng replied:
Yup, this is just to make testing a bit simpler, we can probably add a test case with multiple backends in follow up PR?

aditighag commented on test/k8sT/manifests/lrp-test.yaml:
```diff
@@ -0,0 +1,136 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: k8s1-backend
```
The port names don't match with the ones specified in the yaml, is that intentional? Maybe this is the remote be? On a related note, does it make sense to name the pods as local-backend and remote-backend?

Weil0ng replied:
> The port names don't match with the ones specified in the yaml

I'm a bit confused, which port name are you referring to?
 
> does it make sense to name the pods as local-backend and remote-backend

Since "local" and "remote" are relative to the requester, I think maybe we should keep the name generic here?

Weil0ng replied:
Great catch, done :)

joestringer commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,67 @@ var _ = Describe("K8sServicesTest", func() {
 		})
 	})
 
+	SkipContextIf(func() bool { return !helpers.RunsOnNetNextOr419Kernel() }, "Checks local redirect policy", func() {
```
The [GSG](https://docs.cilium.io/en/latest/gettingstarted/local-redirect-policy/) doesn't mention anything about kernel version, neither does the [Kernel version requirements](https://docs.cilium.io/en/latest/operations/system_requirements/#required-kernel-versions-for-advanced-features). Is there a reason we are restricting the versions we run this test on?

Weil0ng replied:
Quoting @brb "LRP depends on the host-reachable services feature which is enabled only on net-next and 4.19 CI runs."

joestringer replied:
Ok, then might be worth following up to make this clearer in the docs.

aditighag replied:
That's right. I've already updated the documents in the follow-up PR. For the socket lookup call, we need at least kernel version 5.1, but I checked the other day, and the default in GKE seems to be 4.19, which is okay for now.

aditighag replied:
It's worth finding out if we need the host-reachable functionality for the node-local dns case. @Weil0ng Can you revert the kernel check, and trigger CI test suite with `test-me-please`/`retest-4.9`?

aditighag replied:
@brb confirmed that the host-reachable functionality is needed for being able to do service translation from host ns. This would apply for pods deployed in the `hostNetwork` mode. So the minimum kernel version required would indeed be 4.19.

brb replied:
> So the minimum kernel version required would indeed be 4.19.

To be more precise - 4.19.57 for UDP, and 4.17.0 for TCP.

aditighag replied:
Yes, I've kept the version requirement same as kubeproxy-free/host-services features - https://github.com/cilium/cilium/pull/13362/commits/9289cb65f2b0668addb7a9a19758c42cd4419e93

Weil0ng replied:
> It's worth finding out if we need the host-reachable functionality for the node-local dns case.

As @aditighag and @brb pointed out, I think for node-local dns, we need host-reachable services too to capture dns requests from host-ns pods.

joestringer commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,67 @@ var _ = Describe("K8sServicesTest", func() {
+			res := kubectl.ApplyDefault(deploymentYAML)
+			res.ExpectSuccess("Unable to apply %s", deploymentYAML)
+			waitPodsDs()
+			ctx, _ = context.WithTimeout(context.Background(), 1*time.Second)
```
The `cancel` function must be kept & called to clean up context resources.

Weil0ng replied:
Done.

brb reviewed (changes requested):
Some comments.

aditighag commented on test/k8sT/manifests/lrp-test.yaml:
```diff
@@ -0,0 +1,136 @@
+    image: docker.io/cilium/echoserver:1.10.1
+    imagePullPolicy: IfNotPresent
+    ports:
+    - containerPort: 80
```
This won't work, and we need to check if/how CI test is passing. When we have multiple ports, the ports need to be named, and they need to match the ones specified in the pod spec.  The port names are then used to match frontend ports to the corresponding backend ports. 
This is documented in gsg, CRD spec in addition to the code.
@Weil0ng Were you able to run this test successfully?

Weil0ng replied:
Fixed, I can run the test steps manually, but my local e2e test setup is broken (not due to this PR)...

aditighag replied:
Let's get a confirmation from the automated CI run.

aditighag reviewed (changes requested):
Check last comment. Please don't merge until we sort out the issue/get a confirmation that the test is set up correctly.

brb commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,139 @@ var _ = Describe("K8sServicesTest", func() {
+			}
+			for _, tc := range testCases {
+				try := 0
+				for try < repeat {
```
Nit: `for try := 0; try < repeat; try++ { ... }`

Weil0ng replied:
Done.

brb commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,139 @@ var _ = Describe("K8sServicesTest", func() {
+				for try < repeat {
+					res, err := kubectl.ExecInPods(ctx, helpers.DefaultNamespace, tc.selector, curlTCP)
+					Expect(err).To(BeNil(), "http curl failed in %s pod", tc.selector)
+					res[tc.pod].ExpectContains(tc.want)
```
Nit: small comment saying that you are expecting to see pod name in the returned Hostname field would be useful.

Weil0ng replied:
Done.

brb reviewed (approved):
Two nits. Feel free to address them in a follow up PR.

christarazi commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,141 @@ var _ = Describe("K8sServicesTest", func() {
+			res := kubectl.ApplyDefault(deploymentYAML)
+			res.ExpectSuccess("Unable to apply %s", deploymentYAML)
+			pods := []string{feFilter, beFilter}
+			for _, pod := range pods {
```
Nit:

```suggestion
			for _, pod := range []string{feFilter, beFilter} {
```

Weil0ng replied:
Done.

christarazi commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,141 @@ var _ = Describe("K8sServicesTest", func() {
+			pods := []string{feFilter, beFilter}
+			for _, pod := range pods {
+				err := kubectl.WaitforPods(helpers.DefaultNamespace, fmt.Sprintf("-l %s", pod), helpers.HelperTimeout)
+				ExpectWithOffset(1, err).Should(BeNil())
```
I don't think this needs to be an `ExpectWithOffset`. That's reserved for when you have a helper function that's called from inside an `It()` block which performs an assertion. Hence, needing an offset because it's inside a function, rather than in the `It()` itself. This is relevant when the test fails and you get a stacktrace of where the failure occurred.

Weil0ng replied:
Done.

christarazi commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,144 @@ var _ = Describe("K8sServicesTest", func() {
+					}
+				}
+				// Expects to see both backends (local and remote) to appear
+				Expect(seenBe1 && seenBe2).To(BeTrue(), "backend reachable:\n%s\t%v\n%s\t%v", be1Name, seenBe1, be2Name, seenBe2)
```
Maybe we don't need to do a boolean AND here. Probably an individual assertion on both should be sufficient, and removes the need to pretty print the error msg.

Weil0ng replied:
Good point, done :)

christarazi commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,144 @@ var _ = Describe("K8sServicesTest", func() {
+				},
+			}
+			for _, tc := range testCases {
+				for try := 0; try < repeat; try++ {
```
Why do we need to do retries? Is this something that will eventually succeed? If so, then please use `Eventually` for this.

Weil0ng replied:
hmm, not sure if `Eventually` is suitable here. This is to check that lb will "always" send traffic to local backend and "never" send it to remote backend.

christarazi replied:
Ah, I see, so then it's the opposite! We can use `Consistently` to do this :wink:

Weil0ng replied:
makes sense, I think consistently would work for this test, but I doubt either `consistently` or `eventually` matches the semantics in the restore test (line 533), where we need results from multiple curl results to "sometimes" have substring1 and "sometimes" have substring2, any magic work/pattern to use there?

Weil0ng replied:
ah is that just  an `Eventually` && `Eventually` case?

Weil0ng replied:
Done.

christarazi reviewed (changes requested):
A few comments from me regarding testing logic and suggestions to improve / simplify the tests.

christarazi commented on test/k8sT/Services.go:
```diff
@@ -412,6 +412,158 @@ var _ = Describe("K8sServicesTest", func() {
+					res, err := kubectl.ExecInPods(ctx, helpers.DefaultNamespace, tc.selector, tc.cmd)
+					Expect(err).To(BeNil(), "%s failed in %s pod", tc.cmd, tc.selector)
+					returned := res[tc.pod].Stdout()
+					return strings.Contains(returned, tc.want) && strings.Contains(returned, tc.notWant)
```
I think you meant this?
```suggestion
					return strings.Contains(returned, tc.want) && !strings.Contains(returned, tc.notWant)
```

Weil0ng replied:
urghhh, fat fingers :) Fixed.

christarazi reviewed (approved):
LGTM! The use of `Consistently` and `Eventually` simplify the tests nicely. Just one potential typo above.

Weil0ng committed:
Fixes LRP service deletion and restoration

- Deletes all frontends before resetting the config map. Otherwise,
for named ports, later iterations would have a frontend with <nil> IP.

- Moves the restoration logic outside of the deleting loop, since we only
need to restore the service once after *all* frontends with LRP are
removed for this service.

Signed-off-by: Weilong Cui <cuiwl@google.com>

aditighag reviewed (approved):
LGTM. Please add unit tests as a follow-up for the restoration logic to catch the issues that we were debugging yesterday, in case the code path is not covered by the e2e test. For example, assert that service restoration happens only once per policy config when it's deleted.

aditighag reviewed (approved):
We can merge the PR once the LRP test passes.
